Make "make variables" config attributes for overridable flags

levitte opened the pull request:
With the support of "make variables" comes the possibility for the
user to override them.  However, we need to make a difference between
defaults that we use (and that should be overridable by the user) and
flags that are crucial for building OpenSSL (should not be
overridable).

Typically, overridable flags are those setting optimization levels,
warnings levels, that kind of thing, while non-overridable flags are,
for example, macros that indicate aspects of how the config target
should be treated, such as L_ENDIAN and B_ENDIAN.

We do that differentiation by allowing upper case attributes in the
config targets, named exactly like the "make variables" we support,
and reserving the lower case attributes for non-overridable project
flags.

levitte committed:
Make "make variables" config attributes for overridable flags

With the support of "make variables" comes the possibility for the
user to override them.  However, we need to make a difference between
defaults that we use (and that should be overridable by the user) and
flags that are crucial for building OpenSSL (should not be
overridable).

Typically, overridable flags are those setting optimization levels,
warnings levels, that kind of thing, while non-overridable flags are,
for example, macros that indicate aspects of how the config target
should be treated, such as L_ENDIAN and B_ENDIAN.

We do that differentiation by allowing upper case attributes in the
config targets, named exactly like the "make variables" we support,
and reserving the lower case attributes for non-overridable project
flags.

dot-asm commented on Configurations/00-base-templates.conf:
```diff
@@ -116,18 +116,18 @@ my %targets=(
+        MT              => "mt",
+        MTFLAGS         => "-nologo",
+        MTINFLAG        => "-manifest ",
+        MTOUTFLAG       => "-outputresource:",
```
On a **side** note, just for the sake of speaking my mind, \*OUT\* and MT [and eve RC] is the kind of parameterization I find a bit excessive. I'd say they can as well be just coded into windows-makefile.template. Yes, one can argue that there might be some other non-Microsoft toolchain when these would be called something else, but can we actually say that? Template is in fact nmake[!] specific. One can probably argue that we could just as well call it vc-makefile.tmpl. Once again, **side** note, just forget about it :-)

levitte replied:
We've had that discussion before, and while I agree that the template is *nmake* *syntax* specific, I do not agree that it's *vc* specific.

levitte replied:
... or it's at least not meant to be *vc* specific...  there may be details still that make this statement not entirely true, but the intention is there.

dot-asm replied:
Rationale basically is that nmake is something that there is available only with VC, which makes whole thing VC-specific. But of course I can appreciate that templates (or rather unix-Makefile one) are not meant to be tool-chain specific and that it helps to view them *all* as such. However, it's not given that you can escape platform- or tool-chain-specific *rules*. For example, there are mingw-specific rules that use windres in unix-Makefile.template. Currently they disguise as filename-extension-specific, but it's only possible because extension is not used on some other platform. In other words what I'm saying is that I for one wouldn't "feel bad" about having some rules rigid, with no parameterization knobs, sometimes it's appropriate...

levitte replied:
I so strongly disagree with you.  Basically, what you're saying is that if you use nmake, you can only use it with VC.  That's like saying that you only can use GNU make with the GNU C toolchain.

My intention was always that the different templates would be compatible with a particular 'make' implementation for each "platform", but never tied to a particular toolchain (on a best effort basis).  I don't want to change that.

dot-asm replied:
> what you're saying is that if you use nmake, you can only use it with VC.

Trick question... Let's put it this way. If you are using VC, nmake is effectively the only option. [Just in case, using icc for Windows is equivalent to using VC, because it relies *heavily* on remaining infrastructure components.] Other Windows tool-chains do come with own make utilities, which are customarily GNU make. I'm not exactly saying that if you use nmake, you can only use it with VC, but rather that there is nothing else you can use it with. This is just a de-facto statement. But then I also said that I can appreciate that all templates are viewed as tool-chain independent, it *is* useful abstraction. And I also shifted focus toward platform-specific rules. In other words I'm cool with nmake being viewed as tool-chain independent, but it still doesn't have to mean that one has to parameterize down to say MTINFLAG and MTOUTFLAGS. And suggested alternative is to express things like manifest embedding as platform- or tool-chain-specific rules in otherwise tool-chain-neutral template.

dot-asm commented on Configurations/50-masm.conf:
```diff
@@ -11,9 +11,10 @@ my %targets = (
+        AS              => "ml64",
+        ASFLAGS         => "/nologo",
+        ASOUTFLAG       => "/Fo",
+        asflags         => "/c /Cp /Cx /Zi",
```
There is certain contradiction here. Idea is to let user override AS, but if user does that, then it's not given that non-overridable flags fit... One can even wonder if it would be appropriate to have possibility to specify non-overridable assembler. The situation is intensified by the fact that assembler can be associated with perlasm_scheme... Let's say this is a side note too...

levitte replied:
Yeah, I think I agree...  I did hesitate making this change, mind you, and would have no problem changing it back.

levitte replied:
Note that command attributes with names that match "make variables" except for casing are always seen as an alias and are overridable by default.  See [this](https://github.com/openssl/openssl/pull/5534/files#diff-f1206f9fadc5ce41694f69129aecac26R573) and [that](https://github.com/openssl/openssl/pull/5534/files#diff-f1206f9fadc5ce41694f69129aecac26R979) change in Configure.

dot-asm replied:
> Note that command attributes with names that match "make variables" except for casing are always seen as an alias and are overridable by default.

I do understand that. But this doesn't actually lifts contradiction between overridable assembler, non-overridable flags *and* perlasm_scheme. Well, there actually is similar contradiction even for compiler, because there also are non-overridable cflags and overridable CC. So that user is not free to override CC with compiler that doesn't understand non-overridable cflags. But there is no "intensifying factor" such as perlasm_scheme, and there is kind of "tradition" for multiple compilers mimicking most popular one. For example gcc and clang are interchangeable, cl and icc on Windows are interchangeable. But there are no interchangeable options for Windows assembler... Once again, it's side note :-)

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -6,18 +6,21 @@ my $vc_win64a_info = {};
-                                asoutflag => "-o" };
+            $vc_win64a_info = { AS        => "nasm",
+                                ASFLAGS   => "-Ox -g",
+                                asflags   => "-f win64 -DNEAR",
```
Even though -Ox is an optimization flag, it's a bit different with nasm, -Ox is more or less must. Normally switching off optimizations is kind of about debugging, but here your debugging options are not affected a little bit. Though I can argue that nobody will try to override this, Windows users are not into flags that much...

levitte replied:
Will fix.

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -622,9 +630,9 @@ my %targets = (
-        cflags           => combine("-std1 -tune host -fast -readonly_strings",
-                                    threads("-pthread")),
+        CC               => "cc",
+        cflags           => add("-std1 -tune host -fast -readonly_strings",
```
-tune host -fast is an optimization option. Since you keep -O3 as overrideable, this one should be be as well.

levitte replied:
Cool, will do.

(this is precisely why I asked you to review...  you know most of these flags a lot better than I do)

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -857,34 +865,34 @@ my %targets = (
-        cc               => "cl6x",
-        cflags           => "--linux -ea=.s -eo=.o -mv6400+ -o2 -ox -ms -pden",
+        CC               => "cl6x",
+        CFLAGS           => "--linux -ea=.s -eo=.o -mv6400+ -o2 -ox -ms -pden",
```
Only `-o2 -ox -ms` are overrideable, rest is "must".

levitte replied:
Thank you.  Will fix

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -1317,10 +1327,13 @@ my %targets = (
-        defines          => add("OPENSSL_SYS_WIN32", "WIN32_LEAN_AND_MEAN",
+        CC               => "cl",
+        CPP              => '$(CC) /EP /C',
+        CFLAGS           => "-W3 -wd4090 -Gs0 -GF -Gy -nologo",
```
/GF is read-only string pooling. Kind of like -qroconst in aix-cc and there is non-overridable. One can probably say same about all /G options...

levitte replied:
Okie

levitte replied:
(although, as a side note, I was wondering of things like `-qroconst` should be possible to override.  Not that I can understand why anyone would, but...)

dot-asm replied:
Things like -qroconst is rather a declaration of code properties. This one means that there is no code in **our** code base that will try to write to a constant. And without this option you wouldn't be sure. It's also sensible from security perspective, because it kind of ensures that buggy application code can't tamper with our security-sensitive constants. In other words I'd say it appropriate to keep it non-overridable.

dot-asm replied:
On related note one can probably say following. Primary criterion for determining if option should be overridable or not is whether or not option in question **has to be** harmonized with option used to compile application code. Simplest example is debugging options. I mean if you want to debug application, you need debugging library build, right.

levitte replied:
> ... In other words I'd say it appropriate to keep it non-overridable.

Agreed.

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -1498,12 +1510,14 @@ my %targets = (
                                    release => "-O3 -fomit-frame-pointer"),
-        cppflags         => combine("-DL_ENDIAN -DWIN32_LEAN_AND_MEAN",
-                                    "-DUNICODE -D_UNICODE", threads("-D_MT")),
+        CPPFLAGS         => "-DWIN32_LEAN_AND_MEAN",
```
Mmmmm, I'd say WIN32_LEAN_AND_MEAN is required flag. As far as I recall we have type name conflict with some Windows type or variable and WIN32_LEAN_AND_MEAN masks it.

levitte replied:
Ah!  Okie, will fix

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -1624,13 +1640,12 @@ my %targets = (
+        CC               => "cc",
+        CFLAGS           => picker(debug   => "-g -O0",
                                    release => "-O3"),
+        LDFLAGS          => "-Wl,-search_paths_first",
```
I'd say this is required. Without it it will end up trying to link with system libraries, that's not what we need, we want to link with libraries in the build tree.

levitte replied:
Thanks, will fix

dot-asm commented:
CIs are super-angry :-)

levitte commented:
> CIs are super-angry :-)

Yup.  I expected no less, but perhaps not like this...

levitte commented:
Travis is happy now, always something :-)

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -6,18 +6,21 @@ my $vc_win64a_info = {};
+            $vc_win64a_info = { AS        => "nasm",
+                                ASFLAGS   => "-g",
+                                asflags   => "-Ox -f win64 -DNEAR",
+                                ASOUTFLAG => "-o" };
```
This line caught my eye. Even though `-ofile.obj` apparently works, manual page says that it's `-o file.obj`. It might be appropriate to add space...

levitte replied:
Okie

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -29,20 +32,23 @@ sub vc_win32_info {
+                               ASFLAGS   => "",
                                asflags   => "-f win32",
-                               asoutflag => "-o",
+                               ASOUTFLAG => "-o",
```
Same here, i.e. space after `-o`. Just in case, there is space in 1.0.2.

levitte replied:
Will fix

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -1317,9 +1329,12 @@ my %targets = (
+        CC               => "cl",
+        CPP              => '$(CC) /EP /C',
+        CFLAGS           => "-W3 -wd4090 -nologo",
+        COUTFLAG         => "/Fo",
```
Actually with earlier mentioned contradiction between overridable CC and non-overridable cflags, I'd like to argue that COUTFLAG should be non-overridable.

dot-asm replied:
And this applies even to ASOUTFLAG.  [And maybe all \*OUTFLAG?] To reiterate. User is not actually free to override CC or AS with arbitrary value. Replacement has to be flag-compatible with default tool, at least a subset of flags has to be. And what I'm trying to say is that outflags are part of the said subset.

dot-asm commented:
I'd guess appveyor is upset about lack of LEAN_AND_MEAN thing...

dot-asm commented on Configurations/unix-Makefile.tmpl:
```diff
@@ -188,61 +188,111 @@ HTMLSUFFIX=html
+# gcc, then the driver will automatically translate it to -xarch=v8plus
+# and pass it down to assembler.
+AS={- $config{AS} ? "\$(CROSS_COMPILE)$config{AS}" : '$(CC) -c' -}
+ASFLAGS={- join(' ', @{$config{ASFLAGS}}) || '$(CFLAGS)' -}
```
This can't be right. Or more specifically what can't be right is to pass overridable CFLAGS alone to $(CC). One might be able to argue in favour of passing overridable ASFLAGS to overridable $(AS), but not to $(CC). The reason is given in commentary above, there ought to be non-overridable flags present. Well, $(AS) $(ASFLAGS) doesn't seem to be used, at least not on couple of platforms I tested, x86_64 and aarch64. So question might as well be about removal of this, as opposite to fixing it that is.

levitte replied:
It's true that we don't use `$(AS)` or `$(ASFLAGS)` in unix-Makefile.tmpl, we rely on `$(CC)` to do the right thing.  So yeah, I agree that `AS` and `ASFLAGS` can be removed entirely.

levitte replied:
...  what I wonder, thought, is if that's the right thing to do.  Wouldn't *the right thing* be to use `$(AS)` and `$(whatever_ASFLAGS)` when compiling assembler files?  Considering it defaults to `$(CC)` and `$(CFLAGS) `, it will not make a practical difference unless someone actively changes AS and ASFLAGS.

dot-asm replied:
> Wouldn't the right thing be to use $(AS) and $(whatever_ASFLAGS) when compiling assembler files? Considering it defaults to $(CC) and $(CFLAGS),

As already said, one *can* argue in favour of passing overridable flag to overridable tool (so that user bear full responsibility), but it may not default to overridable $(CC) $(CFLAGS) combo. Because we need things like -march, -m64 and alike. Then you mention $(whatever_ASFLAGS), which kind of asks for non-overridable assembler flags. So is suggestion that we would have to maintain non-overridable assember flags for cases when user chooses to override AS and ASFLAGS? [Keep in mind that assemblers tend to have own flags, so you can't just copy-paste few flags from cflags.] In other words *not* using $(AS) is adequate solution, it serves us well so far, while switching to $(AS) arguably raises more questions that we are willing to answer.

levitte replied:
> ... it served us well so far ...

That's enough of an argument for me, and considering we have *never* used them on Unix (sure, there are ASFLAGS assignments all over the Makefiles, but no use of it anywhere...).

dot-asm commented on Configurations/unix-Makefile.tmpl:
```diff
@@ -188,61 +188,111 @@ HTMLSUFFIX=html
+CNF_CFLAGS={- join(' ', $target{cflags} || (),
+                        @{$config{cflags}}) -}
+CNF_CXXFLAGS={- join(' ', $target{cxxflags} || (),
+                          @{$config{cxxflags}}) -}
```
On related note, as problem ought to be somewhere else, when looking at linux-x86_64 file I couldn't help noticing that `CNF_CFLAGS=-pthread -m64`, while `CNF_CXXFLAGS=-std=c++11 -pthread`. I'd say that latter should also have -m64...

levitte replied:
Good point.  That's something to fix in `10-main.conf`

levitte replied:
There *is* another option, *IF* we can say for sure that any C++ flag will simply be an addition to existing C flags, then we can change the semantics of `cxxflags` to be *additional* to `cflags` instead of a replacement.  Your thoughts on this?

Note that it's safe to change the semantics for `cxxflags`, as it's new in 1.1.1-dev.

dot-asm replied:
> Your thoughts on this?

Well, *IF* we can say it for sure, then of course. Trouble is that we can't be 100% sure. It feels that it should work, as long as cxxflags is appended to cflags, but c++ driver doesn't actually have to recognize c-specific options. One example can be -std=c9x... Looking at 10-main modification before last one can start wondering wouldn't be nice if a flag could be *added* to both cflags and cxxflags with one line :-) BTW, default GNU make rules use $(TARGET_ARCH). This kind of gives you an idea that it might be appropriate to introduce similar notion [of common target-specific flag] even for us...

levitte commented:
> I'd guess appveyor is upset about lack of LEAN_AND_MEAN thing...

Hmmm?  I thought I restore that.  Will check

dot-asm commented on Configurations/unix-Makefile.tmpl:
```diff
@@ -188,61 +188,111 @@ HTMLSUFFIX=html
+CNF_CXXFLAGS={- join(' ', $target{cxxflags} || (),
+                          @{$config{cxxflags}}) -}
+CNF_LDFLAGS={- join(' ', $target{lflags} || (),
+                         @{$config{lflags}}) -}
```
Can you double-check if this is right? I'm a bit lost among them, but lflags is documented as "Flags that are used when linking apps.", but CNF_LDFLAGS is passed to LD_LDFLAGS, DSO_LDFLAGS and BIN_LDFLAGS.

levitte replied:
Thing is that `lflags` has been used *everywhere* where linking is involved, DSOs and shared libraries included.  If you look in 1.1.0, you'll see that it's passed in form of `$(LDFLAGS)` in all those cases.  So the docs aren't quite right.

For specialisation, there are other attributes prefixed with `lib_`, `shared_`, `dso_` and `bin_`...  or that's the idea at least.

dot-asm replied:
Maybe it would be appropriate to even rename it as ldflags, because it kind of stings the eye. Everything else is SOMETING = $target{lowercase(SOMETHING)}...

levitte replied:
I agree that it does.  But we're in the middle of a minor release, and I'm trying my best to keep a certain amount of backward compatibility, should anyone have made their own private config files (which we do encourage)...  so I think it's a no can do until 1.2.

(although, I *am* pondering some attribute name translations...  but I'll reserve that for another PR)

levitte commented:
> > I'd guess appveyor is upset about lack of LEAN_AND_MEAN thing...
> 
> Hmmm?  I thought I restore that.  Will check

Ah, I know why, there are things lacking in `windows-makefile.tmpl`, still...  I'm actually going through the build files right now and filling in things I see missing...

levitte committed:
Add space to asoutflag value where appropriate

levitte committed:
Configurations/10-main.conf: Duplicate cflags to cxxflags

dot-asm commented on Configurations/unix-Makefile.tmpl:
```diff
@@ -188,61 +188,131 @@ HTMLSUFFIX=html
+# Variables starting with LIB_ are used to build library object files
+# and shared libraries.
+# Variables starting with DSO_ are used to build DSOs and their object files.
+# Variables starting with BIN_ are used to build programs and their object files.
```
Not that it matters a lot, but this line is 81 characters long...

levitte replied:
YeahIknow...  I was hoping noone would notice.  Ah well ;-)

dot-asm commented:
I get following on for VC-WIN64A:
```
CFLAGS=-W3 -wd4090 -nologo
CNF_CFLAGS=-Gs0 -GF -Gy /MD /O2
```
/O2 belongs in CFLAGS.

levitte commented:
Fixed

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -6,16 +6,19 @@ my $vc_win64a_info = {};
-            $vc_win64a_info = { as        => "ml64",
+            $vc_win64a_info = { AS        => "ml64",
+                                ASFLAGS   => "",
                                 asflags   => "/c /Cp /Cx /Zi",
```
BTW, /Zi is "Add symbolic debug info". You kind of suggest to keep it overridable...

levitte replied:
> You kind of suggest to keep it overridable...

Are you saying it should be overridable?

levitte replied:
Done

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -1498,12 +1530,14 @@ my %targets = (
                                    release => "-O3 -fomit-frame-pointer"),
-        cppflags         => combine("-DL_ENDIAN -DWIN32_LEAN_AND_MEAN",
-                                    "-DUNICODE -D_UNICODE", threads("-D_MT")),
+        cflags           => "-m32",
```
cxxflags even here?

dot-asm replied:
Well, there is no CXX definition, so there is no contradiction... but if you choose to do something about it, there is similar problem in mingw64.

levitte replied:
Like you say, there's no CXX definition, and not in mingw64 either...

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -1624,13 +1660,12 @@ my %targets = (
+        CFLAGS           => picker(debug   => "-g -O0",
                                    release => "-O3"),
         cppflags         => threads("-D_REENTRANT"),
+        ldflags          => "-Wl,-search_paths_first",
```
I suppose this is typo, as we already established this flag should sting the eye as `lflags` :-) ... Yes, attempt to run ./Configure darwin-ppc-cc yields Makfile without -Wl,-search_paths_first

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -1659,15 +1694,16 @@ my %targets = (
     "darwin64-x86_64-cc" => {
         inherit_from     => [ "darwin-common", asm("x86_64_asm") ],
-        cflags           => add("-arch x86_64 -Wall"),
+        CFLAGS           => add("-Wall"),
```
On *side* note this looks misplaced, in sense that it could have appeared in darwin-common. Oh well..

levitte replied:
Yeah...  we do need to clean this up (I've an unfinished branch that merges things like this into toolchain templates that I should make a PR from)

dot-asm commented on Configurations/10-main.conf:
```diff
@@ -1730,8 +1766,9 @@ my %targets = (
-        cflags           => "-mrtp -mhard-float -mstrict-align -fno-implicit-fp -O2 -fstrength-reduce -fno-builtin -fno-strict-aliasing -Wall",
+        CC               => "ccppc",
+        CFLAGS           => "-O2 -Wall",
+        cflags           => "-mrtp -mhard-float -mstrict-align -fno-implicit-fp -fstrength-reduce -fno-builtin -fno-strict-aliasing",
```
One can wonder which of these belong in CFLAGS, but we probably don't care, right? But at least -fstrength-reduce is an optimization option. Shug! Side note.

levitte replied:
Changed

dot-asm commented on Configurations/descrip.mms.tmpl:
```diff
@@ -311,8 +311,8 @@ PERLASM_SCHEME={- $target{perlasm_scheme} -}
-              $x .= "/DEFINE=($d)";
-              $x; -}
+              $x .= "/DEFINE=($d)" if $defines;
+              $x; -} 
```
There is redundant/invisible space at the end of this line. If you care that is. Just in case, I don't really look at descrip.mms.tmpl. This simply caught the eye in git log -p

levitte replied:
Thanks.

dot-asm commented:
`version -f` output lost a bunch of flags, basically those that now listed in CNF_CPPFLAGS and CNF_CFLAGS.

dot-asm commented:
BTW, here are couple of corner cases. What this flag separation is about? Among other things[?] it's about facilitating automated package builds, right? What's so special about them? They kind of like specifying distro-specific CFLAGS, right? So that it's not impossible to imagine that Configure and make receive same CFLAGS.

Corner case 1. `env CFLAGS=-foo-bar ./config -extra-flag` followed by `env CFLAGS=-foo-bar make -e` or `make CFLAGS=-foo-bar`. -extra-flag gets combined with -foo-bar in Makefile, but is not used during compilation. `env CFLAGS=-foo-bar make`, i.e. without -e, uses both flags.

Corner case 2. `./config -extra-flag CFLAGS=-foo-bar`. -extra-flag does not appear in CFLAGS in Makefile (well, I suppose this is kind of expected), but[!] it gets merged into CXXFLAGS. [Yes, in 1st case -extra-flag is added to CXXFLAGS too.]

Presumably one can argue that it's wrong to expect that -extra-flag passed on command line plays well with CFLAGS. I mean one ought to choose only one at a time. But this raises question of conveying the message to user. A note in a README? But there is also kind usability issue. Maybe it wouldn't be inappropriate to *terminate* configuration process if user passes -extra-flag *and* either of recognized variables. i.e. to actually force user to choose...

dot-asm commented on Configurations/unix-Makefile.tmpl:
```diff
@@ -188,60 +188,131 @@ HTMLSUFFIX=html
 # dependent assembler flags. E.g. if you throw -mcpu=ultrasparc at SPARC
 # gcc, then the driver will automatically translate it to -xarch=v8plus
 # and pass it down to assembler.  In any case, we do not define AS or
 # ASFLAGS for this reason.
```
I'd like to suggest to move this comment upwards, closer to CC [maybe even right after CC]. It looks kind of misplaced between RC and RM.

levitte replied:
Sure

levitte commented:
> Among other things[?] it's about facilitating automated package builds, right?

Quite frankly, it's to appease all those who claim (quite correctly, I've found) that the GNU way is near universal, at least in the Unix flavor of things.  I didn't actively think of automated package builds, but it's a good thing they can benefit too...

> Corner case 1. `env CFLAGS=-foo-bar ./config -extra-flag` followed by `env CFLAGS=-foo-bar make -e` or `make CFLAGS=-foo-bar`. -extra-flag gets combined with -foo-bar in Makefile, but is not used during compilation. `env CFLAGS=-foo-bar make`, i.e. without -e, uses both flags.

I'm not sure I understand...  when I do `env CFLAGS=-foo-bar ../master/config -extra-flag`, I get this in the resulting Makefile:

```
CFLAGS=-foo-bar -extra-flag -Wa,--noexecstack
```

Then when you give a new CFLAGS with the `make` run (i.e. `make CFLAGS=-foo-bar`), that new CFLAGS value *replaces* the one in Makefile.  I don't understand how `-extra-flag` gets combined in...

`make` will of course not make the environment priority over the variables on Makefile.  I'm not sure how that's an issue.

> Corner case 2. `./config -extra-flag CFLAGS=-foo-bar`. -extra-flag does not appear in CFLAGS in Makefile (well, I suppose this is kind of expected), but[!] it gets merged into CXXFLAGS. [Yes, in 1st case -extra-flag is added to CXXFLAGS too.]

So yes, when `Configure` finds a flag it doesn't know, it will append it internally to the values known as CFLAGS and CXXFLAGS.  *But*, when it encounters a "make variable" assignment, that variable gets reset to the new value, all previously collected values (such as random flags on the command line) are thrown away.  On the other hand, with `./config CFLAGS=-foo-bar -extra-flag`, you will end up with this in Makefile:

```
CFLAGS=-foo-bar -extra-flag -Wa,--noexecstack
CXXFLAGS=-Wall -O3 -extra-flag -Wa,--noexecstack
```

(note that `-Wa,--noexecstack` is passed on the command line to Configure by `config`)

levitte commented:
> `version -f` output lost a bunch of flags, basically those that now listed in CNF_CPPFLAGS and CNF_CFLAGS.

Fixed...  I hope.

levitte committed:
Configure et al: rename all dso_* to module_* in shared-info.pl

Because there are already attributes with the dso_ prefix that are
used instead of the corresponding lib_ attributes rather than in
addition to them, it gets confusing to have similar or exactly the
same attributes working with different semantics on Unix.

So we rename those by changing the prefix dso_ to module_, and having
those work just like the shared_ attributes, but for DSOs.

levitte commented:
I did a semantic change just now.  In `Configurations/shared-info.pl` and `Configure`, `dso_*` got changed to `module_*`.

The reason for this is that the `dso_` attributes had different semantics depending on the platform flavor we run on.  On Windows and VMS, `dso_lflags` was used *instead of* `lib_lflags`, while on Unix, `dso_lflags` was used *in addition* to `lib_lflags`, much like `shared_ldflag` was for shared libraries.

So `module_cflags` et al are introduced as additional flags to build DSOs, in the same way as how `shared_cflag` et al are used.

This shouldn't disrupt anything, because the `dso_` prefix was practically only used on native Windows and VMS, and `shared-info.pl` is introduced in master, so this is all news for 1.1.1 anyway.

dot-asm commented:
> I'm not sure how that's an issue.

It's confusing and error-prone. Imagine you have a package script that passes extra arguments to config[ure], and did it for years. And now it also picks CFLAGS from environment.

> -Wa,--noexecstack

It should really be detected in ./Configure and go to CNF_CFLAGS.

OK, let me put it as not a question, but suggestion. And suggestion is to refuse to mix the two by failing Configure with corresponding error message.

levitte commented:
> > I'm not sure how that's an issue.
> 
> It's confusing and error-prone. Imagine you have a package script that passes extra arguments to config[ure], and did it for years. And now it also picks CFLAGS from environment.

Good point...  and that brings us to this:

> OK, let me put it as not a question, but suggestion. And suggestion is to refuse to mix the two by failing Configure with corresponding error message.

I agree, that seems to be the safest path.

> > -Wa,--noexecstack
> 
> It should really be detected in ./Configure and go to CNF_CFLAGS.

Errrr...  so if I do `./config -DFOO`, `Configure` will get these options: `-DFOO -Wa,--noexecstack`
How is `Configure` going to know which one is my user input (goes to CFLAGS) and which one is generated from `config` (and could be argued should go to `CNF_CFLAGS`)?

Personally, I think `config` has *no* business generating flags like this, those should be moved elsewhere...  or we should accept that they are going to be treated the same way as any other user generated flag.

levitte commented:
> I agree, that seems to be the safest path.

But before that, we need to resolve the issue with things passed from `config`

dot-asm commented:
> Personally, I think config has no business generating flags like this,

Well, I also something wish that one could just do things without taking all the factors and circumstances into consideration. But we are just not that lucky :-)

> those should be moved elsewhere... or we should accept that they are going to be treated the same way as any other user generated flag.

Or they could be conveyed to Configure in such a way that will make it accept them as non-overridable. I also wish we could just skip it, but it might be inappropriate, not at this point. As for -Wa,--nonexecstack, it's kind of a security option, and it's argued that Configure is more appropriate place to detect it, because it covers cross-compile scenarios. It's argued that it's too easy to miss...

levitte commented:
Well, I've got an idea, coming up soon in a new commit here

levitte committed:
config: Pass diverse flags to Configure via the environment

Passing flags "discovered" by 'config' on the command line to
'Configure' mixes them up with flags given by the user.  That is
contrary to their intention, so they need to be passed in a different
manner.

Enter the environment variables __CNF_CPPDEFINES, __CNF_CPPINCLUDES,
__CNF_CPPFLAGS, __CNF_CFLAGS, __CNF_CXXFLAGS, __CNF_LDFLAGS, and
__CNF_LDLIBS, initialised by 'config', and then used by Configure to
initialise these %config values: defines, includes, cppflags, cflags,
cxxflags, lflags, and ex_libs.

levitte commented:
There...

dot-asm commented on config:
```diff
@@ -519,9 +530,12 @@ case "$GUESSOS" in
+	    EV56|PCA56)		__CNF_CFLAGS="$__CNF_CFLAGS -mcpu=ev56"
+                                __CNF_CXXFLAGS="$__CNF_CXXFLAGS -mcpu=ev56";;
+	    *)			__CNF_CFLAGS="$__CNF_CFLAGS -mcpu=ev6"
+                                __CNF_CXXFLAGS="$__CNF_CXXFLAGS -mcpu=ev6";;
```
It's inconsistent mix of tabs and spaces here is in multiple spaces. In diff it looks like jagged indentation.

levitte replied:
Will fix

dot-asm commented on Configure:
```diff
@@ -1276,8 +1298,8 @@ if ($disabled{pic})
                     shared_defines shared_includes shared_ldflag
-		    dso_cflags dso_cxxflags dso_cppflags
-                    dso_defines dso_includes dso_lflags))
+		    module_cflags module_cxxflags module_cppflags
```
Inconsistent mix of tabs and spaces even here.

levitte replied:
Will fix

dot-asm commented on Configurations/windows-makefile.tmpl:
```diff
@@ -495,7 +587,7 @@ EOF
          return <<"EOF";
 $obj$objext: $deps
-	\$(AS) \$(ASFLAGS) \$(ASOUTFLAG)\$\@ $srcs
+	\$(AS) $asflags \$(ASOUTFLAG)\$\@ $srcs
```
A bit of an overkill, but all right. I mean I can't imagine varying assembler flags...

dot-asm commented on Configurations/unix-Makefile.tmpl:
```diff
@@ -1024,7 +1100,7 @@ EOF
 	\$(CC) \$(LIB_CFLAGS) $linkflags\$(LIB_LDFLAGS)$shared_soname$shared_imp \\
 		-o $target_full$shared_def $objs \\
-                \$(PLIB_LDFLAGS) $linklibs \$(EX_LIBS)
+                \$(PLIB_LDFLAGS) $linklibs \$(LIB_EX_LIBS)
```
I don't see that PLIB_LDFLAGS is assigned anywhere... And there are no plib_lfags left in Configurations/*.conf. Note that it was -Wl,-search_paths_first and original idea behind plib was place it in specific place on command line... In other words this needs double-checking on MacOS X, and in worst case changes would need to be reverted...

levitte replied:
Ah, right, need to clean that too.  It should get passed via `LIB_LDFLAGS`, so `PLIB_LDFLAGS` has become irrelevant,  I do suspect that the exact position of `-Wl,-search_paths_first` isn't crucial as long as it happens before all the -L and -l arguments...

levitte committed:
fixup! config: Pass diverse flags to Configure via the environment

levitte committed:
fixup! Configure et al: rename all dso_* to module_* in shared-info.pl

levitte committed:
Configure et al: get rid of last traces of plib_lflags / PLIB_LDFLAGS

levitte committed:
Configurations/README: update documentation on flags

dot-asm commented:
So we kind of agreed that Configure should reject attempts to pass additional flags in case any of recognized "make variables" are set/passed, forcing user to make the choice among the two. Remaining question is here and now or later.

levitte commented:
> Remaining question is here and now or later.

For clarity, I prefer to make that a separate PR

dot-asm commented:
What does "UNFINISHED" in last commit mean? That more commits are still to come?

levitte commented:
... actually, mind you, the behavior with mixed compiler options and variable assignments is documented.  Read [INSTALL](https://github.com/openssl/openssl/blob/master/INSTALL#L584-L606)

levitte commented:
> What does "UNFINISHED" in last commit mean? That more commits are still to come?

Ah, yes!  Thanks for the reminder.  I need to finish documenting, that's what it means.

dot-asm commented:
> ... actually, mind you, the behavior with mixed compiler options and variable assignments is documented

The fact the we have this request does mean that outcome is not quite right, and it needs re-thinking. And documentation should naturally be part of this rethinking. Well, I probably wouldn't find mixing additional flags and "make variables" inappropriate *if* latter were not picked from environment at config time, e.g. as `env CFLAGS=-foo-bar ./config -bar-foo`. In other words as long as picking environment is an option, I'd feel strong about prohibit mixing. Well, one can argue in favour of not looking at environment variables if user passes *any* additional flags... Either way there should be something that makes it less error-prone... It's just that prohibiting mixing is easiest to explain, and would probably be easy for user to accept.

levitte committed:
fixup! Configure et al: rename all dso_* to module_* in shared-info.pl

dot-asm commented:
> one can argue in favour of not looking at environment variables if user passes any additional flags...

Actually on second thought this might be as sensible. I mean instead of prohibiting mixing, one can disregard environment variables if any tool-chain flags are passed at config command line. This will ensure "backward compatibility". Mixing stuff on command line still doesn't feel quite right, but I suppose some would appreciate it...

dot-asm commented on Configurations/50-masm.conf:
```diff
@@ -11,9 +11,10 @@ my %targets = (
+        AS              => "ml64",
+        ASFLAGS         => "/nologo",
         asoutflag       => "/Fo",
+        asflags         => "/c /Cp /Cx /Zi",
```
We have moved /Zi to ASFLAGS in 10-main.conf. It would be appropriate to follow the lead...

levitte replied:
Fixed

levitte commented:
> Actually on second thought this might be as sensible. I mean instead of prohibiting mixing, one can disregard environment variables if any tool-chain flags are passed at config command line. This will ensure "backward compatibility". Mixing stuff on command line still doesn't feel quite right, but I suppose some would appreciate it...

No, because that would mean that these two lines would produce different results:

```
env CFLAGS=-foo ./config -bar

env ./config CFLAGS=-foo -bar
```

That would be utterly inappropriate and would be quite a surprise.

levitte committed:
fixup! Make "make variables" config attributes for overridable flags

dot-asm commented on Configurations/README:
```diff
@@ -45,19 +45,24 @@ In each table entry, the following keys are significant:
                            Note: if the same feature is both enabled
                            and disabled, disable wins.
 
+        as              => The Assembler compiler.  This is not always
```
Wow! I've never thought of assembler as "Assembler compiler". Take `man as` for example. Assembler is effectively opposed to compiler... As in "compiler generates input for assembler"... If you try to pull analogy with C compiler, then word before compiler is supposed to refer to language, while assembler as term refers rather to program...

dot-asm replied:
Hmmm, wouldn't it be appropriate to list it as AS now? Along with CC, CPP, etc...

levitte replied:
Well, therein lies the double meaning, doesn't it?  "Assembler" refers to the language as well as the program that translates it into an object file.  Now, this gets even more complicated with MACRO32 on Alpha, where VAX assembler is translated (compiled) into Alpha assembler, and it really is refered to as an "assembler compiler".  But that's a rather unusual case, so I'll just remove "compiler"...

levitte replied:
That some C compilers translate C to assembler is quite irrelevant, actually.

levitte replied:
Fixed

dot-asm commented on Configurations/README:
```diff
@@ -67,28 +72,40 @@ In each table entry, the following keys are significant:
-        plib_lflags     => Extra linking flags to appear just before
-                           the libraries on the command line.
+        lflags          => Default flags used when linking apps,
+                           shared libraries or DSOs [4].
```
Why not just "Default linker flags [used each time linker is invoked]"? Side note...

levitte replied:
Good point.  I changed a few more accordingly.

dot-asm commented:
> No

Then prohibition it is.

levitte committed:
Configure: disallow the mixture of compiling flags and env / make variables

Note that this might give surprising results if someone forgets an environment
variable that has been set previously.

levitte committed:
fixup! Configurations/README: update documentation on flags

levitte committed:
fixup! Configurations/README: update documentation on flags

dot-asm commented on Configure:
```diff
@@ -875,6 +891,17 @@ while (@argvcopy)
+        die <<"_____";
+***** Mixing env / make variables and configure line compiling or linking flags
+***** is not permitted.
+***** Affected env / make variables: $detected_env
```
Well, I'm probably not exactly in position to judge, but I find the message a bit too cryptic to unravel. Maybe "additional compiler/linker flags as command line option" would be better than "configure line compiling or linking flags"?

On side note I tried env CFLAGS=-foo-bar ./config -lfoo, and it said that affected variable is LDLIBS. It's per se correct, but maybe it would also be useful to print offending variable, one that triggered the condition.

levitte replied:
Good point about the message, will fix.

Regarding "offending", you do realise that CFLAGS is hardly offending for a flag that ends up in LDLIBS, right? ;-) (there's a reason I chose "affected") 
The only way to fix that is to display the name of all involved env vars that are set as well. That's a simple fix, will do.

dot-asm replied:
What I'm trying to say is if I imagine myself being a user and face the error message in question, in first hand I'd appreciate to see what is it in my environment that triggered the failure. Given the message I'd start looking for environment variable called LDLIBS, and if I don't have one, I'd be inclined to conclude that error message is not helpful.

dot-asm replied:
> The only way to fix that is to display the name of all involved env vars that are set as well.

Yes, that would do.

levitte replied:
> What I'm trying to say is if I imagine myself being a user and face the error message in question,

Yup, I got what you meant

levitte replied:
Fixed

levitte committed:
fixup! Configure: disallow the mixture of compiling flags and env / make variables

levitte committed:
fixup! Configure: disallow the mixture of compiling flags and env / make variables

dot-asm reviewed (approved):
Red cross from Travis is unrelated.

levitte commented:
Thank you!

Merged. 48dcca265c9cb0c26459240ab1858577dacf9e67 to b92013602b6666314ad200d48e10a1aaa2c5b2fa

dot-asm commented:
> > what you're saying is that if you use nmake, you can only use it with VC.

> Trick question ... I'm not exactly saying that if you use nmake, you can only use it with VC, but rather that there is nothing else you can use it with. This is just a de-facto statement.

Well, on closer look it wouldn't be wrong to say that if you use windows-makefile.tmpl, you can only use it with VC. Because there is VC-specific command sequence, one that generates dependency list with /Zs /showIncludes output being piped to perl script that expects quite specific format. Original side note closed with "one can probably argue that we could just as well call it vc-makefile.tmpl". I'd say that one can still do that... And it's still a side note :-) :-) :-)

levitte commented:
You do know that all you're doing is to tempt me into coming up with new config attributes, right?  ;-)

richsalz commented:
Maybe this is good enough for now, and we should focus on other PR reviews?
